File: Medium/17.Letter_Combinations_of_a_Phone_Number.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []

        res = [""]
        d = {'2': "abc", '3': "def", '4': "ghi", '5': "jkl", 
            '6': "mno", '7': "pqrs", '8': "tuv", '9': "wxyz"}

        for digit in digits:
            news = []
            for s in res:
                for ch in d[digit]:
                    news.append(s + ch)
            res = news

        return res

File: Java/Medium/17.Letter_Combinations_of_a_Phone_Number.java
package Medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

class Solution {

    static final String[] letters = new String[] {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) 
            return new LinkedList<>();
        
        LinkedList<String> ans = new LinkedList<>();
        ans.add("");
        
        for (int i = 0; i < digits.length(); i++) {
            int size = ans.size();
            for (int j = 0; j < size; j ++) {
                String prevString = ans.removeFirst();
                for (char letter : letters[digits.charAt(i) - '0'].toCharArray()) { 
                    ans.add(prevString + letter);
                }
            }
        }
        return ans;
    }
}